hydroflow/tests/surface_syntax.rs

```diff
@@ -0,0 +1,45 @@
+use hydroflow::hydroflow_parser;
+
+#[test]
+pub fn test_parser_basic() {
+    // hydroflow_parser! {
+    //     edges_input = (input() ->);
+
+    //     init_vertex = (seed([0]) ->);
+    //     // loop_vertices = (->);
+    //     out_vertices = (-> for_each(|x| println!("Reached: {}", x)));
+
+    //     reached_vertices = (merge[init_vertex, loop_vertices] -> map(|v| (v, ())));
+
+    //     (join[reached_vertices, edges_input] -> map(|(_src, ((), dst))| dst) -> dedup() -> tee[loop_vertices, out_vertices]);
+
+    //     // x = (a -> b() -> c() -> (a -> b -> c) -> p);
+    //     // b = (a -> b() -> c() -> (a -> b -> c) -> p);
+    //     // x = (a -> b() -> c() -> (a -> b -> c) -> p);
+    //     // x = (a -> b() -> c() -> (a -> b -> c) -> p);
+    // }
+
+    hydroflow_parser! {
+        edges_input = input();
+        init_vertex = seed([0]);
+        out_vertices = for_each(|x| println!("Reached: {}", x));
+
+        // reached_vertices = (merge[init_vertex, loop_vertices] -> map(|v| (v, ())));
+        reached_vertices = (merge() -> map(|v| (v, ())));
+        (init_vertex -> [0]reached_vertices);
+
+
+        map_dedup = (map(|(_src, ((), dst))| dst) -> dedup());
+        my_join = (join() -> map_dedup);
+        (reached_vertices -> [0]my_join);
+        (edges_input -> [1]my_join);
+
+        my_join_tee = (my_join -> tee());
```

Comment:
I'm surprised that tee needs to be explicitly invoked, shouldn't it be implicit when we detect a pull-to-push?

Reply:
So yeah, there were a few options, one option is to have tee-ing and merging (anti-tee) being implied.

So that would be:
```rs
my_join -> [1]reached_vertices
my_join -> out_vertices
```

But when we have operators that actually have multiple inputs and outputs it can create some ambiguity. Like if you want to merge two streams before joining
```rs
j = join()
a -> j[0]
b -> j[0]
c -> j[1]
```
I suppose could work, but it could be confusing. So decided to make it explicit for joins and merges.

Reply:
Yeah, explicit merges make sense (and correspond to FP APIs in languages with unrestricted aliasing). I guess for tee-ing I am more used to implicit because you can just perform multiple maps on the same iterator reference in some languages.

A middle ground that could potentially work would be to have teeing as a special language feature where you explicitly declare every place you want to tee out a node? Similar to using `clone` on Rc I guess.

So like:
```rust
my_join.tee -> [1]reached_vertices
my_join.tee -> out_vertices
```

which makes things a bit less awkward when you want to tee to many other nodes